subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/junit4/JUnit4MultiVersionTest.groovy

```diff
@@ -28,10 +29,17 @@ trait JUnit4MultiVersionTest extends JUnit4CommonTestSources {
         String configureTestFramework = "useJUnit()"
 
         @Override
-        String getTestFrameworkDependencies() {
-            return """
-                testImplementation 'junit:junit:${MultiVersionIntegrationSpec.version}'
-            """
+        String getTestFrameworkDependencies(String sourceSet) {
+            if (MultiVersionIntegrationSpec.version.startsWith("3")) {
```

Comment:
It would be nice to have a comment explaining why this adds a 3 and 4 dependency -- it's not immediately clear why this is done to me.

Reply:
Done.  With test suites, the runtime dependency wouldn't be necessary because we would inject the runtime dependency when you call `useJUnit()` on the suite.  The runtime dependency is only necessary when not using test suites.  I think previously you could just add the JUnit 3 version to `testImplementation` and you just happened to also get Gradle's version of JUnit 4 on the classpath at runtime, but since we deprecated these implicit dependencies (to fix other classpath pollution problems) you now have to specify both in order to avoid a deprecation warning (i.e. it still works, but you get a deprecation warning and then the test fails, so for testing we specify both).

This makes me realize there's another problem, though.  If you were using test suites but specified a JUnit 3 version in the suite (i.e. `useJUnit('3.8.2')`) it looks like you end up with only the JUnit 3 library at runtime and it fails.  It might be ok to only support JUnit4 in test suites, as I'm not sure there is a good use case where you _must_ use JUnit 3 to compile, but we should at least throw an error here if you specify a non-4.x version.